GH 9016: Bitwise operation weirdness

tvyomkesh opened the pull request:
Series now supports bitwise op for integral types. 

I have made the changes in wrapper() itself instead of na_op() because it looked to me like wrapper is controlling the input and output fill and dtype. Once that is taken care of, na_op() seems to be doing the right thing by itself and so I did not have to change anything to get back the expected behavior. Happy to change if things need to be altered toward better design etc.

tvyomkesh committed:
GH 9016: Bitwise operation weirdness

Series now supports bitwise op for integral types

tvyomkesh commented:
This PR intends to resolve #9016

shoyer commented on pandas/core/ops.py:
```diff
@@ -657,17 +658,31 @@ def wrapper(self, other):
-            other = other.reindex_like(self).fillna(False).astype(bool)
-            return self._constructor(na_op(self.values, other.values),
+            if other.dtype.kind not in ['u', 'i']:
+                other = other.reindex_like(self).fillna(False).astype(bool)
```
the `other.reindex_like(self)` bit is repeated twice here -- can you consolidate it?

shoyer commented on pandas/core/ops.py:
```diff
@@ -657,17 +658,31 @@ def wrapper(self, other):
+                other = other.reindex_like(self).fillna(0).astype(other.dtype)
+
+            if self.dtype.kind in ['u', 'i'] and other.dtype.kind in ['u', 'i']:
+                return self._constructor(na_op(self.values, other.values),
```
same concern about repetition here with the duplicated `self._constructor` call

shoyer commented on pandas/core/ops.py:
```diff
@@ -657,17 +658,31 @@ def wrapper(self, other):
-                                    index=self.index).fillna(False)
-            return res.astype(bool).__finalize__(self)
+            if isinstance(other, numbers.Integral) and self.dtype.kind in ['u', 'i']:
+                res = self._constructor(na_op(self.values, other),
```
and as above

jreback commented:
can u move the tests near the other boolean tests (right below) in test_series

also need some test for ops with NaN and series containing nan and ops with other dtypes
(with other dtypes should prob raise TypeError)

tvyomkesh commented:
@shoyer, @jreback  thanks, I am working on the suggested changes. In the meantime I noticed this. (using master without this patch).
```
In [4]: Series([0,1,2,3]) & Series([True, True, True, True])
Out[4]:
0    False
1     True
2    False
3     True
dtype: bool
```
I was expecting output to be `Series([False, True, True, True])` as RHS has dtype bool. 
Figured Numpy converts bool to 0/1 and executes bitwise op. For e.g.,
```
In [5]: import numpy as np

In [6]: np.array([0,1,2,3]) & np.array([True, True, True, True])
Out[6]: array([0, 1, 0, 1])
```
In the Series example, should the behavior be preserved? In that case should the output dtype be bool or int64? Otherwise it looks confusing.

tvyomkesh commented:
Looks like if changed, `test_logical_with_nas()` in test_frame.py would break.

tvyomkesh commented:
made the following changes:
* Code refactored to removed some duplication
* Moved test function near other comparison tests
* Added more tests with NaN and other dtypes
* Fixed code so it works with list,tuple and np.ndarray types correctly.

jreback commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -107,6 +107,8 @@ Enhancements
 - ``Timedelta`` will now accept nanoseconds keyword in constructor (:issue:`9273`)
 
+- ``Series`` now supports bitwise operation for integral types (:issue:`9016`)
+
```
add a mini-example here

jreback commented on pandas/core/ops.py:
```diff
@@ -657,17 +658,46 @@ def wrapper(self, other):
-            other = other.reindex_like(self).fillna(False).astype(bool)
-            return self._constructor(na_op(self.values, other.values),
+            other = other.reindex_like(self)
+            if other.dtype.kind not in ['u', 'i']:
```
look in ``core.common`` for function like ``is_interger_dtype`` and such. Pls use these rather then checking dtypes like this

jreback commented:
@tvyomkesh this seems like a lot of branching for just slightly different dtype handling. see what you can do about that.

tvyomkesh commented:
@jreback thanks for pointing to ``core.common.is_interger_dtype()``. Made the changes. Also cut down on branching a bit. Updated release note with a small example.

I still think this is confusing.
```
In [3]: pd.Series([2]) & pd.Series([True])
Out[3]:
0    False
dtype: bool
```
Either both the inputs should be converted to dtype boolean in which case output value would be True  or if both the inputs are converted to dtype int64 implictly (and bitwise & is carried out) then the output dtype should remain int64 and not be coerced to bool. The current patch can be modified to fix this. Thoughts?

tvyomkesh commented:
PR review changes are done. Good to merge?

jreback commented on pandas/core/ops.py:
```diff
@@ -654,20 +654,48 @@ def na_op(x, y):
+            is_other_int_dtype = com.is_integer_dtype(other.dtype)
+
+            if is_other_int_dtype:
+                other = other.fillna(0)
```
you can collapse a lot of this if you define functions like this:

```
fill_bool = lambda x: x.fillna(False).astype(bool)
fill_int = lamda x: x.fillna(0)

filler = fill_int if is_self_int_dtype and is_other_int_dtype else fill_bool

other = fill(other)
res = fillter(self._constructor(.......))
```

tvyomkesh replied:
done

jreback commented on pandas/core/ops.py:
```diff
@@ -654,20 +654,48 @@ def na_op(x, y):
+                                    index=self.index)
+
+            if isinstance(other, (list, tuple, np.ndarray)) and \
+                all(com.is_integer(x) for x in other) and is_self_int_dtype:
```
instead of doing this list-comp just do (you can incorporate in what I show above as well)

```
if isinstance(other, (list,tuple,np.ndarray)) and is_self_int_dtype:
    if com.is_integer_dtype(np.array(other,copy=False)):
        res = res.fillna(0)
```

tvyomkesh replied:
done

tvyomkesh commented:
@jreback thanks. I have made the changes.

jreback commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -117,6 +117,20 @@ Enhancements
+
+- ``Series`` now supports bitwise operation for integral types (:issue:`9016`)
+  .. code-block:: python
+  In [2]: pd.Series([0,1,2,3], list('abcd')) | pd.Series([4,4,4,4], list('abcd'))
```
indent 2 spaces here (the code under the directive), otherwise won't render correctly

tvyomkesh replied:
oh didn't notice this one. thanks.

jreback commented on pandas/core/ops.py:
```diff
@@ -654,20 +654,33 @@ def na_op(x, y):
-            return res.astype(bool).__finalize__(self)
+            # scalars, list, tuple, np.array
+            is_other_int_dtype = True if com.is_integer_dtype(np.array(other,copy=False)) else False
+
```
you don't need the if else
is_integer_dtype already returns a boolean

shoyer replied:
also, `np.asarray(values)` is a succinct alias for `np.array(values, copy=False)`

tvyomkesh replied:
@jreback yes, should have assigned the test value directly like in line 665. Not sure what I was thinking here :)
@shoyer neat, I will change to use asarray.

tvyomkesh commented:
@jreback and @shoyer I have pushed the changes. Thanks.

jreback commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -117,6 +117,21 @@ Enhancements
 - ``StringMethods.pad()`` and ``center()`` now accept `fillchar` option to specify filling character (:issue:`9352`)
 
+
+
```
ok move this to the api section snd show s code block of the existing behavior (and thrn out this after)

tvyomkesh replied:
done.

jreback commented:
thanks @tvyomkesh
